documentloaders/cloudsql/documentloader.go

```diff
@@ -0,0 +1,364 @@
+		return ""
+	}
+	writer.Flush()
+	return strings.TrimSuffix(sb.String(), "\n") // Remove trailing newline
+}
+
+// yamlFormatter formats row data into a YAML string.
+func yamlFormatter(row map[string]any, contentColumns []string) string {
+	var sb strings.Builder
+	for _, column := range contentColumns {
+		if val, ok := row[column]; ok {
+			sb.WriteString(fmt.Sprintf("%s: %v\n", column, val))
+		}
+	}
+	return strings.TrimSpace(sb.String())
+}
+
+// jsonFormatter formats row data into a JSON string.
+func jsonFormatter(row map[string]any, contentColumns []string) string {
+	data := make(map[string]any)
+	for _, column := range contentColumns {
+		if val, ok := row[column]; ok {
+			data[column] = val
+		}
+	}
+	jsonData, err := json.Marshal(data)
+	if err != nil {
+		// Should not happen in normal cases as values are usually simple types
+		return ""
+	}
+	return string(jsonData)
+}
+
+// parseDocFromRow parses a Document from a row of data.
+func (l *DocumentLoader) parseDocFromRow(row map[string]any) (schema.Document, error) {
+	pageContent := l.formatter(row, l.contentColumns)
+	metadata := make(map[string]any)
+
+	populateJSONMetadata := func(data []byte) error {
+		var jsonMetadata map[string]any
+		err := json.Unmarshal(data, &jsonMetadata)
+		if err != nil {
+			return fmt.Errorf("failed to parse JSON from column '%s': %w", l.metadataJSONColumn, err)
+		}
+		for k, v := range jsonMetadata {
+			metadata[k] = v
+		}
+		return nil
+	}
+
+	if l.metadataJSONColumn != "" {
+		value := row[l.metadataJSONColumn]
+		switch v := value.(type) {
+		case []byte:
+			if err := populateJSONMetadata(v); err != nil {
+				return schema.Document{}, err
+			}
+		case string:
+			if err := populateJSONMetadata([]byte(v)); err != nil {
+				return schema.Document{}, err
+			}
+		default:
+			return schema.Document{}, fmt.Errorf("failed to parse JSON from column '%s': invalid column type", l.metadataJSONColumn)
+		}
+	}
+
+	for _, column := range l.metadataColumns {
+		if column != l.metadataJSONColumn {
+			metadata[column] = row[column]
+		}
+	}
+
+	return schema.Document{
+		PageContent: pageContent,
+		Metadata:    metadata,
+	}, nil
+}
+
+// Load executes the configured SQL query and returns a list of Document objects.
+func (l *DocumentLoader) Load(ctx context.Context) ([]schema.Document, error) {
+	rows, err := l.engine.Pool.Query(ctx, l.query)
+	if err != nil {
+		return nil, fmt.Errorf("failed to execute query: %w", err)
+	}
+	defer rows.Close()
+
+	fieldDescriptions := rows.FieldDescriptions()
+	columnNames := make([]string, len(fieldDescriptions))
+	valuesPrt := make([]any, len(columnNames))
+
+	for i, fd := range fieldDescriptions {
+		columnNames[i] = fd.Name
+		t, ok := l.oidProcessor.process(pgtype.OID(fd.DataTypeOID))
+		if !ok {
+			valuesPrt[i] = new(sql.RawBytes)
+		}
+		valuesPrt[i] = t
+	}
+
+	var documents []schema.Document
+	for rows.Next() {
+		if err := rows.Scan(valuesPrt...); err != nil {
```

Comment:
We shouldn't need the processors. We should be able to use the type from the scan and not introduce the dep `database/sql`. We may have to pass in the field names or struct.
@kurtisvg  to review.

Reply:
@davidnastasi I'm not sure I understand the problem the processors are trying to solves. @davidnastasi can you provide some additional details or context on why we need them?

Reply:
Should be able to do something like:

```
	results, err := t.Pool.Query(ctx, t.Statement)
	if err != nil {
		return nil, fmt.Errorf("unable to execute query: %w", err)
	}

	fields := results.FieldDescriptions()

	var out []any
	for results.Next() {
		v, err := results.Values()
		if err != nil {
			return nil, fmt.Errorf("unable to parse row: %w", err)
		}
		vMap := make(map[string]any)
		for i, f := range fields {
			vMap[f.Name] = v[i]
		}
		out = append(out, vMap)
	}
```